src/plugins/visualizations/public/embeddable/create_vis_embeddable_from_object.ts

```diff
@@ -69,8 +70,14 @@ export const createVisEmbeddableFromObject = (deps: VisualizeEmbeddableFactoryDe
       return new DisabledLabEmbeddable(vis.title, input);
     }
 
-    const indexPattern = vis.data.indexPattern;
-    const indexPatterns = indexPattern ? [indexPattern] : [];
+    let indexPatterns: IndexPattern[] = [];
+
+    if (vis.type.getUsedIndexPattern) {
+      indexPatterns = await vis.type.getUsedIndexPattern(vis.params);
+    } else if (vis.data.indexPattern) {
+      indexPatterns = [vis.data.indexPattern];
+    }
```

Comment:
One idea that came to mind:

I'm wondering if, instead of having two sources of truth in a vis for the index pattern (either `vis.data.indexPattern` OR `vis.type.getUsedIndexPattern`), it would be cleaner to make a change in `vis` itself so that `vis.data.indexPattern` always reads from `vis.type.getUsedIndexPattern`.

In `vis.ts` we could do something like:
```ts
this.data.indexPattern = await this.type.getIndexPattern(this.data, this.params);
```

Then on the vis type:
```ts
await getIndexPattern(data: VisData, params: VisParams) {
  ...etc
}
```

Then on the vis type, `getIndexPattern` is optional, and defaults to the logic that's currently in `vis.ts`:
https://github.com/elastic/kibana/blob/a17dcada46866b3ffc242bd7feda809c88e4e250/src/plugins/visualizations/public/vis.ts#L140-L153

WDYT @alexwizp? Do you think that would work / make sense?

Reply:
Pretty the same idea was implemented in my initial commit(https://github.com/elastic/kibana/pull/82964/commits/e2e02fe70280b1a32b7ee9d6c068753fa6c44526), but after discussing it @ppisljar we decided to don't modify `vis.setState` method.